.github/workflows/build_test.sh

```diff
@@ -100,11 +100,11 @@ add-apt-repository -y ppa:upstream-systemd-ci/systemd-ci
 apt-get -y update
 apt-get -y build-dep systemd
 apt-get -y install "${PACKAGES[@]}"
-# Install the latest meson and ninja form pip, since the distro versions don't
-# support all the features we need (like --optimization=). Since the build-dep
+# Install more or less recent meson and ninja with pip, since the distro versions don't
+# always support all the features we need (like --optimization=). Since the build-dep
 # command above installs the distro versions, let's install the pip ones just
 # locally and add the local bin directory to the $PATH.
```

Comment:
Pfff, we seem to be doing this backwards. We do not want to depend on the latest meson, we're not building a CI for meson. We should depend on a reasonably-recent stable version of meson and other tools. Distros are doing the packaging of meson so that we don't have to.

Also, the hashes are terrible. Tools have versions so that *humans* can understand the tool version. Bi-weekly update of those hashes is just ugly noise in the logs.

Dunno, if the distro is doing things wrong, let's just pick a different distro. I know that Fedora builds meson in all supported releases within a few days after an upstream release, and the update becomes available after passing through automated and non-automated CI, usually a week or so later. Using a package makes to me much more sense than building everything from upstream git hashes and creating ever more complexity.

Reply:
> We do not want to depend on the latest meson

Can't argue with that but somehow the systemd project tends to rely on features that aren't available in the versions of Ubuntu it's tested on. That's where all those `pip install`s came from.

> the hashes are terrible

I agree they are. But those hashes will be updated automatically by Dependabot (by analogy with https://github.com/evverx/systemd/pull/58) so I wouldn't worry about that.

> Bi-weekly update of those hashes is just ugly noise in the logs.

I don't think `meson` is released that often.

> if the distro is doing things wrong, let's just pick a different distro

In theory it should be possible to run a VM of some kind on GitHub Actions but I don't think it's possible to replace the Docker image provided by OSS-Fuzz. Either way, it seems to be much complicated than what's propose here I think.

Reply:
Dropped `pip` in https://github.com/systemd/systemd/pull/21459. Let's see how it goes. It's worth mentioning that OSS-Fuzz will probably keep their ubuntu image up until EOL so it would be hard to rely on more or less recent features there without pip

Reply:
It didn't work out apparently
```
meson compile -C build
usage: meson [-h]
             {setup,configure,dist,install,introspect,init,test,wrap,subprojects,help,rewrite}
             ...
meson: error: unrecognized arguments: -C build
+ fatal ''\''meson compile'\'' failed with --optimization=0'
+ echo -e '\033[31;1m'\''meson compile'\'' failed with --optimization=0\033[0m'
'meson compile' failed with --optimization=0
+ exit 1
```

Reply:
I'd add that this PR is supposed to prevent systemd from being `meson` CI by switching from one stable version to another under control

Reply:
Looking at https://pypi.org/project/meson/#history It would probably make sense to switch from "weekly" to "monthly".

Reply:
Switched to "monthly"

Reply:
by the way, judging by https://github.com/mesonbuild/meson/issues/9585 `0.60.1` seems to have broken "-Db_coverage" so it would probably nice to be able to jump from version to version on CentOS CI as well

Reply:
That doesn't affect Linux apparently luckily  :-)

Reply:
Speaking of other distributons, @mrc0mmand I wonder if packit-as-a-service can be used to run the "build" workflow with `lto`, `--optimization` and so on? As far as I understand it uses Fedora under the hood somewhere

Reply:
Not yet, unfortunately. There's still one blocker left (which should be fixed soon, though).

Reply:
@mrc0mmand good to know. Thanks! If would be great if it could be configurable in the systemd repository somehow to make it easier to affect those builds. As far as I can see, If I wanted to change something I'd have to edit the Fedora package (which isn't ideal I think).

> building everything from upstream git hashes and creating ever more complexity

FWIW those hashes have nothing to with upstream git hashes. They are computed based on stable releases like https://pypi.org/project/meson/0.60.1/#files.

Reply:
In the meanitime, I switched to `apt-get` on OSS-Fuzz (since it seems to be working there). I'm not sure 0.53.2 can be considered reasonably recent for the next four years (as far as I understand its EOL is in 2025) but that's what we get depending on packages provided by distributions :-)

Reply:
Looking at https://github.com/mesonbuild/meson/issues/9300, I think another advantage of this PR is that when meson rolls out their backwards incompatible updates as they usually do it shouldn't affect "stable" branches with meson pinned to specific versions that should just keep working without backporting changes like e92777d275a7f8f05 everywhere (assuming systemd-stable is supposed to be tested as well of course)

Reply:
And looking at https://github.com/systemd/systemd-stable/commit/cdc1cd4eb3f86f3100d6d04b1b5d12a2d87b8704, it seems some part of testing infrastructure is already pinned so I guess it seems to be a good example why it can be necessary sometimes to pin some dependencies in `systemd-stable` to get the tests to run there

Reply:
I've just pushed another commit where I pinned meson in `.lgtm.yml`. Looking at https://github.com/systemd/systemd/commit/55d651d8f69919b98cdc062e312e4454c34428ef, https://github.com/systemd/systemd/commit/7ad1f043984386a8bb56a4a556942dcb02b1fd84 and https://github.com/systemd/systemd/commit/8f876748d3f89d822b1167214ed4fb66146dc3df I think that's another example of why meson has to be installed with `pip` sometimes and it can certainly benefit from this PR. I'm waiting for https://lgtm.com/projects/g/systemd/systemd/rev/pr-e650a6e5701c6ebc6fafcf0a1326b2ddfc3582c9 to pass to make sure I didn't break anything.

Reply:
@keszybz to just kind of sum up and somewhat address your concerns

> if the distro is doing things wrong, let's just pick a different distro

I'm all for switching to Fedora (or any other distribution shipping more or less recent versions of meson) but unfortunately it isn't possible on OSS-Fuzz and LGTM where we have to deal with what they use under the hood

> building everything from upstream git hashes

Those hashes have nothing to do with upstream git hashes. They are computed with `pip hash` based on stable meson releases and used to verify the integrity of packages installed by `pip`. The versions (like `0.60.1`) are included in `requirements.txt` as well.

I think at the end of the day it boils down to whether the people maintaining the CI (that would be @mrc0mmand mostly :-)) find this change useful and can live with it.

Reply:
As for me, I'd go and merge this, since as @evverx pointed out some meson pinning in CI is inevitable (unfortunately).

Reply:
@keszybz Dependabot caught two broken updates in https://github.com/systemd/systemd/pull/21505 (mkosi v11) and https://github.com/systemd/systemd/pull/21340 (super-linter 4.8.3) . It would be great if it could also catch meson updates like that. My understanding is that at this point this PR has two ACKs and @topimiettinen seems to have liked the hashes :-) It would be great if it could get an ACK from you.

Reply:
GitHub recommends pinning 3rd party actions to a full length SHA: 
https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions

Pinning Meson to a version and SHA seems to be a good idea, instead of trusting that `pip` can't ever install meson from a compromised site. I'm not sure I get all nuances here though. Checking GPG signatures would give me much warmer feelings, but then the trusted keys would need to be stored here and kept up to date (OK, now we'd keep SHAs up to date).

Reply:
There's also action to check that actions use only SHAs ([ensure-sha-pinned-actions](https://github.com/marketplace/actions/ensure-sha-pinned-actions)).

Reply:
...turtles all the way down!

Reply:
I set up pinning for actions and meson in my new project [libaslrmalloc](https://github.com/topimiettinen/libaslrmalloc) and added ensure-sha-pinned-actions for extra turtles. It seems to work as advertised. Likewise for [ld-so-daemon](https://github.com/topimiettinen/ld-so-daemon) but it's OK with ancient ubuntu meson, so no need for pip pinning.

Reply:
> Checking GPG signatures would give me much warmer feelings, but then the trusted keys would need to be stored here and kept up to date

I don't think PYPI can enforce GPG signatures at this point. To be fair, considering that `systemd` itself doesn't sign release tarballs I wouldn't expect all those python packages to be signed either :-) All those signatures magically appear when packages are installed via `apt-get`, `dnf` and so on but we can't always use them unfortunately.

> There's also action to check that actions use only SHAs (ensure-sha-pinned-actions).

Good to know. Thanks! It can't be used here in its current form though because CIFuzz can't be pinned properly.

Reply:
GPG signatures also consist of SHA hash, but it's cryptographically signed with a key. Trusting GPG signatures means that the SHAs don't matter, only the key which signed them. In that model, dependabot should only accept SHAs which are signed by a trusted key (which one?).

I suppose `apt` could be told not to use keys provided by the OS but only the set of keys given by the build script. Also `apt-get install` could be broken down to steps of:
1. Package file download
2. SHA check and pinning
3. Crypto checks
4. Package installation from file

I don't know if that would be useful. If the OS on GitHub couldn't be trusted, a lot of things aren't very trustworthy either.

Reply:
It's also a bit strange that GitHub dependency graph says that meson comes from `jsoref/meson` which is a fork of `mesonbuild/meson`. Perhaps SHA matches also on forks and dependabot doesn't know which is the official repo. The checks still use pypi.org:
```
updater | INFO <job_236618121> Starting job processing
updater | INFO <job_236618121> Starting update job for topimiettinen/libaslrmalloc
updater | INFO <job_236618121> Checking if meson 0.60.1 needs updating
  proxy | 2021/11/24 22:12:44 [014] GET https://pypi.org:443/simple/meson/
  proxy | 2021/11/24 22:12:44 [014] 200 https://pypi.org:443/simple/meson/
updater | INFO <job_236618121> Latest version is 0.60.1
updater | INFO <job_236618121> No update needed for meson 0.60.1
```

Reply:
@topimiettinen I agree. What I was trying to say is that PYPI has never been supposed to deal with signing or verifying releases published there or anything like that (because as far as I know the idea was to make it much easier to use it and we all know that GPG is far from that :-) so in a way all those hashes is kind of an afterthought I think that nevertheless should help to mitigate the "curl | bash" nature of `pip install` so to speak.

FWIW `cosign` is supposed to make all that easier. It even supports "keyless" mode: https://github.com/sigstore/cosign/blob/main/KEYLESS.md (it still uses keys of course but that part is somewhat hidden) but it isn't production-ready.

> It's also a bit strange that GitHub dependency graph says that meson comes from jsoref/meson which is a fork of mesonbuild/meson.

I have to admit I'm not sure what it is. Could you point me in the right direction?

Reply:
I found it at https://github.com/topimiettinen/libaslrmalloc/network/dependencies. That's wrong of course but I don't think it has anything to do with Dependabot version updates. Dependabot uses PYPI and never points to that fork (or any other repository including the real meson repository for that matter)

Reply:
Looking at https://github.com/mesonbuild/meson/network/dependents and https://github.com/jsoref/meson/network/dependents I'm pretty sure it's a GitHub bug. It could be `jsoref/meson` just turned on the dependencies graph before `mesonbuild/meson` did.

Reply:
I reported it in https://github.com/mesonbuild/meson/issues/9634.

Reply:
Now that I'm on a mission to fix the GitHub dependency graph apparently :-) I can confirm that that GitHub bug has nothing to do with Dependabot version updates used in this PR. Dependabot always uses PYPI (unless other registries are specified explicitly). I'm not sure what else I can do here :-)

Reply:
The dependency graph was fixed by the way :-)

Reply:
Good, it was a bit creepy.